code/game/objects/items/weapons/scrolls.dm

```diff
@@ -16,92 +16,60 @@
+	. += "<span class='notice'>P.S. Don't forget to bring your gear, you'll need it to cast most spells.</span>"
 
-/obj/item/teleportation_scroll/Topic(href, href_list)
-	..()
-	if(usr.stat || usr.restrained() || src.loc != usr)
+/obj/item/teleportation_scroll/attack_self(mob/user)
+	if(!uses) //somehow?
+		to_chat(user, "<span class='warning'>You attempt to read the scroll but it disintegrates in your hand, it appears that is has run out of charges!</span>")
+		qdel(src)
 		return
-	var/mob/living/carbon/human/H = usr
-	if(!( istype(H, /mob/living/carbon/human)))
-		return 1
-	if((usr == src.loc || (in_range(src, usr) && istype(src.loc, /turf))))
-		usr.set_machine(src)
-		if(href_list["spell_teleport"])
-			if(src.uses >= 1)
-				teleportscroll(H)
-	attack_self(H)
-	return
 
-/obj/item/teleportation_scroll/proc/teleportscroll(mob/user)
-
-	var/A
-
-	A = input(user, "Area to jump to", "BOOYEA", A) as null|anything in SSmapping.teleportlocs
-
-	if(!A)
+	var/picked_area
+	picked_area = input(user, "Area to jump to", "Teleport where?", picked_area) as null|anything in SSmapping.teleportlocs
+	if(!picked_area)
 		return
 
-	var/area/thearea = SSmapping.teleportlocs[A]
-
+	var/area/thearea = SSmapping.teleportlocs[picked_area]
 	if(user.stat || user.restrained())
 		return
-	if(!((user == loc || (in_range(src, user) && istype(src.loc, /turf)))))
-		return
 
-	if(thearea.tele_proof && !istype(thearea, /area/wizard_station))
+	if(!(user == loc || (in_range(src, user) && isturf(user.loc))))
+		return //They can't use it if they put it in their bag or drop it and walk off, but if they are stood next to it they can.
+
+	if(thearea.tele_proof && !istype(thearea, /area/wizard_station)) //Nowhere in SSmapping.teleportlocs should be tele_proof, but better safe than sorry
 		to_chat(user, "<span class='warning'>A mysterious force disrupts your arcane spell matrix, and you remain where you are.</span>")
 		return
 
 	var/datum/effect_system/smoke_spread/smoke = new
-	smoke.set_up(5, 0, user.loc)
+	smoke.set_up(5, 0, get_turf(user))
 	smoke.attach(user)
 	smoke.start()
 	var/list/L = list()
+
 	for(var/turf/T in get_area_turfs(thearea.type))
-		if(!T.density)
-			var/clear = 1
-			for(var/obj/O in T)
-				if(O.density)
-					clear = 0
-					break
-			if(clear)
-				L+=T
+		if(is_blocked_turf(T))
+			continue
+		L.Add(T)
 
-	if(!L.len)
+	if(!length(L))
 		to_chat(user, "<span class='warning'>The spell matrix was unable to locate a suitable teleport destination for an unknown reason. Sorry.</span>")
 		return
 
 	if(user && user.buckled)
 		user.buckled.unbuckle_mob(user, force = TRUE)
 
-	if(user && user.has_buckled_mobs())
+	if(user?.has_buckled_mobs())
 		user.unbuckle_all_mobs(force = TRUE)
 
-	var/list/tempL = L
-	var/attempt = null
-	var/success = FALSE
-	while(tempL.len)
-		attempt = pick(tempL)
-		user.forceMove(attempt)
-		if(get_turf(user) == attempt)
-			success = TRUE
-			break
-		tempL.Remove(attempt)
+	user.forceMove(pick(L))
+	smoke.start()
```

Comment:
isnt this the same smoke at the location of your teleport? 
dont you need to make new smoke at L?

Reply:
It works because of `smoke.attach(user)`

So smokes at both locations.